File: data/poller.go
// Copyright (c) 2026 lazybeeper by Ronen Druker.

package data

import (
	"time"

	tea "charm.land/bubbletea/v2"
)

// Default polling intervals.
const (
	// ChatPollInterval is how often the chat list is refreshed.
	ChatPollInterval = 5 * time.Second
	// MessagePollInterval is how often the active chat messages are refreshed.
	MessagePollInterval = 3 * time.Second
	// IdleBackoffInterval is the polling interval when no changes are detected.
	IdleBackoffInterval = 15 * time.Second
	// IdleTimeout is how long without changes before polling backs off.
	IdleTimeout = 30 * time.Second
)

// TickMsg signals that a polling tick has fired.
type TickMsg struct {
	// Kind distinguishes chat-list ticks from message ticks.
	Kind TickKind
}

// TickKind identifies the type of polling tick.
type TickKind int

const (
	// ChatTick polls the chat list.
	ChatTick TickKind = iota
	// MessageTick polls the active chat's messages.
	MessageTick
)

// Poller manages polling state and idle backoff logic.
type Poller struct {
	// lastChatChange tracks when the chat list last changed.
	lastChatChange time.Time
	// lastMsgChange tracks when messages last changed.
	lastMsgChange time.Time
	// lastChatCount tracks the previous chat count for diff detection.
	lastChatCount int
	// lastMsgCount tracks the previous message count for diff detection.
	lastMsgCount int
	// enabled controls whether polling is active.
	enabled bool
}

// NewPoller creates a new Poller in the enabled state.
func NewPoller() *Poller {
	now := time.Now()

	return &Poller{
		lastChatChange: now,
		lastMsgChange:  now,
		enabled:        true,
	}
}

// SetEnabled turns polling on or off.
func (p *Poller) SetEnabled(enabled bool) {
	p.enabled = enabled
}

// NotifyChatChange records that the chat list has changed.
func (p *Poller) NotifyChatChange(count int) {
	if count != p.lastChatCount {
		p.lastChatChange = time.Now()
		p.lastChatCount = count
	}
}

// NotifyMsgChange records that the message list has changed.
func (p *Poller) NotifyMsgChange(count int) {
	if count != p.lastMsgCount {
		p.lastMsgChange = time.Now()
		p.lastMsgCount = count
	}
}

// ChatTickCmd returns a tea.Cmd for the next chat poll, using idle backoff.
func (p *Poller) ChatTickCmd() tea.Cmd {
	if !p.enabled {
		return nil
	}

	interval := ChatPollInterval
	if time.Since(p.lastChatChange) > IdleTimeout {
		interval = IdleBackoffInterval
	}

	return tea.Tick(interval, func(time.Time) tea.Msg {
		return TickMsg{Kind: ChatTick}
	})
}

// MessageTickCmd returns a tea.Cmd for the next message poll, using idle backoff.
func (p *Poller) MessageTickCmd() tea.Cmd {
	if !p.enabled {
		return nil
	}

	interval := MessagePollInterval
	if time.Since(p.lastMsgChange) > IdleTimeout {
		interval = IdleBackoffInterval
	}

	return tea.Tick(interval, func(time.Time) tea.Msg {
		return TickMsg{Kind: MessageTick}
	})
}

File: src/data/poller.ts
// Copyright (c) 2026 lazybeeper by Ronen Druker.

/** How often the chat list is refreshed (ms). */
export const CHAT_POLL_INTERVAL = 5_000;

/** How often the active chat messages are refreshed (ms). */
export const MESSAGE_POLL_INTERVAL = 3_000;

/** Polling interval when no changes are detected (ms). */
export const IDLE_BACKOFF_INTERVAL = 15_000;

/** How long without changes before polling backs off (ms). */
export const IDLE_TIMEOUT = 30_000;

/** Identifies the type of polling tick. */
export enum TickKind {
  /** Polls the chat list. */
  ChatTick = "chat",
  /** Polls the active chat's messages. */
  MessageTick = "message",
}

/**
 * Poller manages polling state and idle backoff logic.
 */
export class Poller {
  private lastChatChange: number;
  private lastMsgChange: number;
  private lastChatCount = 0;
  private lastMsgCount = 0;
  private enabled = true;

  constructor() {
    const now = Date.now();
    this.lastChatChange = now;
    this.lastMsgChange = now;
  }

  /**
   * Turns polling on or off.
   * @param enabled - Whether polling should be enabled.
   */
  setEnabled(enabled: boolean): void {
    this.enabled = enabled;
  }

  /**
   * Returns whether polling is currently enabled.
   * @returns True if polling is enabled.
   */
  isEnabled(): boolean {
    return this.enabled;
  }

  /**
   * Records that the chat list has changed.
   * @param count - The current number of chats.
   */
  notifyChatChange(count: number): void {
    if (count !== this.lastChatCount) {
      this.lastChatChange = Date.now();
      this.lastChatCount = count;
    }
  }

  /**
   * Records that the message list has changed.
   * @param count - The current number of messages.
   */
  notifyMsgChange(count: number): void {
    if (count !== this.lastMsgCount) {
      this.lastMsgChange = Date.now();
      this.lastMsgCount = count;
    }
  }

  /**
   * Returns the appropriate chat poll interval, using idle backoff.
   * @returns The poll interval in milliseconds.
   */
  chatInterval(): number {
    if (!this.enabled) {
      return 0;
    }

    if (Date.now() - this.lastChatChange > IDLE_TIMEOUT) {
      return IDLE_BACKOFF_INTERVAL;
    }

    return CHAT_POLL_INTERVAL;
  }

  /**
   * Returns the appropriate message poll interval, using idle backoff.
   * @returns The poll interval in milliseconds.
   */
  messageInterval(): number {
    if (!this.enabled) {
      return 0;
    }

    if (Date.now() - this.lastMsgChange > IDLE_TIMEOUT) {
      return IDLE_BACKOFF_INTERVAL;
    }

    return MESSAGE_POLL_INTERVAL;
  }
}
